### PR TITLE
Introduce SPORES in v0.7.0 as a generalisable mode

### PR DESCRIPTION
**Pull request by FLomb**
## Summary of changes in this pull request

* Introduces a `spores` scenario as part of the national-scale example model
* Introduces a custom math file in the national-scale example model, which is activated in the `spores` scenario to create a slack cost constraint
* It provides a dedicated notebook to generate SPORES iteratively based on the `spores` scenario overrides and a custom Python function for assigning penalties at each iteration.

Based on internal discussions, we have come to realise that there are many ongoing SPORES developments that may be hard to always reconduct to a monolithic run mode in the core code. Instead, it may be best to implement SPORES as a method that simply allows you to run a model in `plan` mode once, and then N number more times with some external function being applied between each run, with the user providing that function. This way, this mode could be used not only for SPORES but also for robustness analysis and other things. 

This PR aims to show how SPORES can be indeed already used in v0.7.0 based on default software functionality plus a simple for-loop and a custom Python function to assign penalties for decision variables at each iteration of the loop. To move from this to the generalised mode described above, we would only need to decide how to make some bits of this code (such as the number of iterations, and the custom Python function) a default request for inputs to the user.

The new `run-iteratively-with-a-function` mode (a better name must be found, could also be `spores-and-more` if the main application remains SPORES), would essentially require the following three key inputs: 1) how many iterations to run; 2) what function to apply at each iteration; and 3) what backend parameter to updated based on the function.

## Reviewer checklist

- [ ] Test(s) added to cover contribution
- [ ] Documentation updated
- [ ] Changelog updated
- [ ] Coverage maintained or improved

### CODE REVIEW

**Comment by FLomb** [comment]
I see that building the docs fails because I am using matplotlib, which isn't part of the default Calliope environment. I will remove the plotting part, for now. We can later re-implement it via a supported library, or just leave it out.

**Comment by brynpickering** [comment]
@FLomb I've updated the approach to minimise the input requirements from the user. The main change is that there is now no longer a spores-specific cost class. Instead, the objective is changed in the math. Updating this objective with user-defined math would allow a user to update how the spores scores are translated to penalties in the optimisation problem.

The SPORES mode stuff is currently untested although you can see that it works in the example notebook.

I hope that some of the hardcoding of params in `model.py` can go away with the implementation of #712. 

The SPORES mode stuff will also move out of `model.py` (initial work on that is also in #712). Hopefully, though, you can see that it is on the verge of being generalisable - there is an "iterative optimisation" controller that calls a model updater. My idea would be to scrub "spores" from this controller and just have it so that it can run with a supplied updater function, the SPORES mode one being one of them. This would allow for other iterative updaters (including custom SPORES algorithms) to be supplied by the user and managed by the controller.

**Comment by brynpickering** [comment]
@irm-codebase & @FLomb if you could check this PR that would be great!

File: src/calliope/backend/pyomo_backend_model.py [hunk 1]
```diff
@@ -201,6 +203,16 @@ def _objective_setter(
...
+    def set_objective(self, name: str) -> None:  # noqa: D102, override
+        for obj_name, obj in self.objectives.items():
+            if obj.item().active and obj_name != name:
+                self.log("objectives", obj_name, "Objective deactivated.", level="info")
+                obj.item().deactivate()
+            if obj_name == name:
+                obj.item().activate()
+                self.log("objectives", obj_name, "Objective activated.", level="info")
+        self.objective = name
+
```
Comment (irm-codebase) [inline src/calliope/backend/pyomo_backend_model.py:215]: I'd consider directly querying the requested name and using sets to deactivate the rest to reduce condition checks and to 'fail fast'.

Less checks, but same strength.

```suggestion
    def set_objective(self, name: str) -> None:  # noqa: D102, override
        self.objectives[name].activate()
        self.objective = name
        self.log("objectives", name, "Objective activated.", level="info")
        for key in self.objectives.keys() - {name}:
            if self.objectives[key].active:
                self.objectives[key].deactivate()
                self.log("objectives", key, "Objective deactivated.", level="info")
```

File: src/calliope/model.py [hunk 2]
```diff
@@ -266,6 +267,16 @@ def build(
...
+        elif mode == "spores":
+            backend_input = self._model_data.copy()
+            if "spores_score" not in backend_input:
+                backend_input["spores_score"] = xr.DataArray(0).assign_attrs(
+                    is_result=0
+                )
+            if "spores_baseline_cost" not in backend_input:
+                backend_input["spores_baseline_cost"] = xr.DataArray(
+                    np.inf
+                ).assign_attrs(is_result=0)
```
Comment (irm-codebase) [inline src/calliope/model.py:279]: I do not like the idea of hidden values being set here...
Shouldn't we just fail in this case instead?

Comment (irm-codebase) [reply src/calliope/model.py:279]: Perhaps this is something that the schema should take care of.

File: src/calliope/model.py [hunk 3]
```diff
@@ -491,7 +504,7 @@ def _prepare_operate_mode_inputs(
 
         return sliced_inputs
 
-    def _solve_operate(self, **solver_config) -> xr.Dataset:
+    def _solve_operate(self, solver_config: config.Solve) -> xr.Dataset:
         """Solve in operate (i.e. dispatch) mode.
 
         Optimisation is undertaken iteratively for slices of the timeseries, with
```
Comment (irm-codebase) [inline src/calliope/model.py:510]: This is missing a description for `solver_config`.

File: src/calliope/model.py [hunk 4]
```diff
@@ -506,7 +519,9 @@ def _solve_operate(self, **solver_config) -> xr.Dataset:
 
         LOGGER.info("Optimisation model | Running first time window.")
 
-        step_results = self.backend._solve(warmstart=False, **solver_config)
+        step_results = self.backend._solve(
+            warmstart=False, **solver_config.model_dump()
+        )
```
Comment (irm-codebase) [inline src/calliope/model.py:524]: Why use `model_dump`? Shouldn't we just pass `solver_config` directly?

File: src/calliope/model.py [hunk 5]
```diff
@@ -571,3 +588,103 @@ def _recalculate_storage_initial(self, results: xr.Dataset) -> xr.DataArray:
...
+            xr.Dataset: Results dataset.
+        """
+        LOGGER.info("Optimisation model | Resetting SPORES parameters.")
+        self.backend.update_parameter(
+            "spores_score", self.inputs.get("spores_score", xr.DataArray(0))
+        )
+        self.backend.update_parameter(
+            "spores_baseline_cost",
+            self.inputs.get("spores_baseline_cost", xr.DataArray(np.inf)),
+        )
```
Comment (irm-codebase) [inline src/calliope/model.py:608]: Nice to have: 
A `for` might be more maintainable in the long run... Both do the same, after all.

File: src/calliope/model.py
Comment (irm-codebase) [inline src/calliope/model.py:1]: As a general comment: we might want to start moving 'modes' functions out of `model.py` and into a separate subdirectionry in the future to make this file easier to maintain.

File: src/calliope/model.py [hunk 6]
```diff
@@ -571,3 +588,103 @@ def _recalculate_storage_initial(self, results: xr.Dataset) -> xr.DataArray:
...
+                warmstart=False, **solver_config.model_dump()
+            )
+            results_list.append(step_results)
+
+            if spores_config.save_per_spore_path is not None:
+                LOGGER.info(f"Optimisation model | Saving SPORE {spore} to file.")
+                step_results.assign_coords(spores=spore).to_netcdf(
+                    spores_config.save_per_spore_path / f"spore_{spore}.nc"
+                )
+
```
Comment (irm-codebase) [inline src/calliope/model.py:644]: This part is a bit hard to follow. Maybe consider adding a short comment to explain what `results_list` is doing here to someone unfamiliar with the SPORES approach...

File: src/calliope/model.py [hunk 7]
```diff
@@ -571,3 +588,103 @@ def _recalculate_storage_initial(self, results: xr.Dataset) -> xr.DataArray:
...
+                spores_config.tracking_parameter, xr.DataArray(True)
+            ).notnull()
+            & self.inputs.definition_matrix
+        )
+
+        # Look at capacity deployment in the previous iteration
+        previous_cap = previous_results["flow_cap"].where(spores_techs)
+
+        # Make sure that penalties are applied only to non-negligible deployments of capacity
+        min_relevant_size = 0.1 * previous_cap.max(["nodes", "techs"])
```
Comment (irm-codebase) [inline src/calliope/model.py:675]: `0.1` is a 'magic' value. Shouldn't it be in the spores configuration?

**Review by irm-codebase — COMMENTED** [review]
Looks good overall, although a bit hard to follow in places. Added a few suggestions.

File: src/calliope/model.py (hunk 4 above)
Comment (brynpickering) [reply src/calliope/model.py:524]: yeah there's no reason why not tbh. It was using `model_dump` to avoid excessive changes in the PR that introduced the solver config

File: src/calliope/model.py
Comment (brynpickering) [reply src/calliope/model.py:1]: Yeah, I have that implemented on a different feature branch...

File: src/calliope/model.py (hunk 7 above)
Comment (brynpickering) [reply src/calliope/model.py:675]: probably, yes.

File: src/calliope/config.py [hunk 8]
```diff
@@ -215,6 +215,13 @@ class SolveSpores(ConfigBaseModel):
     tracking_parameter: str | None = None
     """If given, an input parameter name with which to filter technologies for consideration in SPORES scoring."""
 
+    score_iteration_threshold_relative: float = Field(default=0.1, ge=0)
+    """A factor to apply to flow capacities above which they will increment the SPORES score.
```
Comment (irm-codebase) [inline src/calliope/config.py:219]: Should the minimum be `gt` (greater than) rather than `ge` (greater or equal)?

File: src/calliope/model.py
Comment (irm-codebase) [reply src/calliope/model.py:1]: Makes sense. For later, then!

File: src/calliope/config.py (hunk 8 above)
Comment (irm-codebase) [reply src/calliope/config.py:219]: Also, I find the name slightly confusing... I'd include the 'unit' explicitly and shorten it a bit.
Would `score_threshold_factor` work?

**Review by irm-codebase — APPROVED** [review]
Changes look good.

I am tentatively approving, with some suggestions on naming.

File: src/calliope/config.py (hunk 8 above)
Comment (brynpickering) [reply src/calliope/config.py:219]: I think zero is a reasonable value to assign to it (i.e., there is no threshold). I'll rename!

**Comment by brynpickering** [comment]
@irm-codebase I've updated the implementation slightly to get the SPORES maths showing in the docs (you can see it under "pre-defined math" in the PR build). It requires an update to our math docs to show all defined objectives by default. To differentiate between those objectives I'm now adding a suffix to the objective name to say whether or not it is activated. Question is, is this confusing for e.g. the `plan` mode math (that it has `(active)` after the objective name)?

**Review by irm-codebase — APPROVED** [review]
@brynpickering changes look fine, but I am not sure why the code difference is showing changes from `main`...

Approved, but let's be careful just in case to avoid overwriting things on  `main`

**Comment by brynpickering** [comment]
> but I am not sure why the code difference is showing changes from main.

Which changes is it showing?

**Comment by irm-codebase** [comment]
@brynpickering ignore, I was incorrect (just goes to show that you should do reviews while well-rested).
For some reason I confused the `sense` in the math with the pydantic PR, but that is not the case.

File: src/calliope/schemas/config_schema.py [hunk 9]
```diff
@@ -134,45 +133,34 @@ class SolveSpores(CalliopeBaseModel):
...
-                raise ValueError("`save_per_spore_path` must be a directory.")
-        return self
+    tracking_parameter: str | None = None
+    """If given, an input parameter name with which to filter technologies for consideration in SPORES scoring."""
+
+    score_threhsold_factor: float = Field(default=0.1, ge=0)
+    """A factor to apply to flow capacities above which they will increment the SPORES score.
+    E.g., if the previous iteration flow capacity was `100` then, with a threshold value of 0.1,
+    only capacities above `10` in the current iteration will cause the SPORES score to increase for that technology at that node.
+    If, say, the current iteration's capacity is `8` then the SPORES score will not change for that technology (as if it had no )
```
Comment (FLomb) [inline src/calliope/schemas/config_schema.py:152]: Something missing at the end of this sentence?

File: docs/examples/modes.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+# Here, we compare installed capacities between the baseline run (== `plan` mode) and the SPORES.
+# Note how the baseline SPORE is the same as `plan` mode and then results deviate considerably.
+
+# %%
+fig_flows_plan = plot_capacity(model_plan.results)
+fig_flows_plan.update_layout(title="Plan mode capacities")
+
+# %%
+fig_flows_spores = plot_capacity(model_spores.results, facet_col="spores")
+fig_flows_spores.update_layout(title="SPORES mode capacities")
```
Comment (FLomb) [inline docs/examples/modes.py:190]: Where is this shown in the published docs? I cannot find it.

File: src/calliope/math/spores.yaml [hunk 11]
```diff
@@ -2,9 +2,37 @@ import:
   - plan.yaml
 
 constraints:
-  cost_sum_max:
+  total_system_cost_max:
+    description: >
+      Limit total system cost. Conceived for use in SPORES mode to apply a maximum
+      relaxation to the system cost compared to the least-cost feasible option.
+    foreach: [costs]
```
Comment (FLomb) [inline src/calliope/math/spores.yaml:9]: Is it normal for this to iterate over all types of costs? In the docs, we say the default is 'monetary', but this doesn't seem to reflect it. Or is it intended that it is open to all costs here, but 'by default' users tend to then specify only one type of cost in their use of this?

**Review by FLomb — COMMENTED** [review]
Most things look in order, apart from a couple of minor comments I have made just to double-check. 

However, I am not quite following how the suggested implementation relates to the idea of a generalisable/open version of the mode. It looks very specific to the actual (old) SPORES method, including hard-coded scoring logics in `model.py`. What am I missing?

File: src/calliope/math/spores.yaml (hunk 11 above)
Comment (brynpickering) [reply src/calliope/math/spores.yaml:9]: True, currently it would apply a constraint limiting all cost classes to `least_cost_val * (1 + spores_slack)`  (note now that the SPORES score _isn't_ a cost). I suppose this would matter if we were tracking other costs in the model (e.g. CO2 emissions). We could limit this to `monetary` in the math.

**Comment by brynpickering** [comment]
> However, I am not quite following how the suggested implementation relates to the idea of a generalisable/open version of the mode. It looks very specific to the actual (old) SPORES method, including hard-coded scoring logics in model.py. What am I missing?

You're right that it hasn't yet been well generalised. I've mentioned what might need to happen to make it possible in [this comment](https://github.com/calliope-project/calliope/pull/716#issuecomment-2561335028).

What I could imagine is that we have "plugins" like our math helper function subclass functionality. That is, there would be a `CalliopeMode` abstract base class that can be subclassed to create custom modes (and from which `operate` and `spores` mode would be defined). It could have methods like `pre_build`, `post_build`, `pre_solve`, and `solve` which would allow a user to define what (if anything) should happen at each of these steps. And then they link to it in their config when defining their mode. However, it's beyond the scope of this PR right now. It will have to wait until post-0.7 unless we can get more resources to deliver it.

File: docs/examples/modes.py (hunk 10 above)
Comment (FLomb) [reply docs/examples/modes.py:190]: Ok, but this part was completely detached from the SPORES docs, making it unlikely to find when looking at SPORES. I've made a commit to facilitate the connection.

**Review by FLomb — COMMENTED** [review]
The functionality looks good as a basic implementation of SPORES. However, it' still outdated and less powerful compared to what we implemented for the study ['What is redundant and what is not'](https://doi.org/10.1016/j.apenergy.2023.121002), and available on a [branch of mine of v0.6.8](https://github.com/FLomb/calliope/blob/0.6.8-turbo-spores-methods/calliope/backend/run.py). 

Given that the idea of having SPORES as a highly customisable mode has been postponed, I think it would make sense for v0.7.0 to at least implement the additional scoring methods as a default option, since many partners have provided feedback and empirical that they are sometimes (depending on the model specifics) critically needed.

File: src/calliope/model.py [hunk 12]
```diff
@@ -567,3 +574,181 @@ def _recalculate_storage_initial(self, results: xr.Dataset) -> xr.DataArray:
...
+                .clip(min=1, max=1)
+                .fillna(0)
+                .where(spores_techs)
+            )
+            return new_score
+
+        def _score_relative_deployment() -> xr.DataArray:
+            """Relative deployment scoring algorithm."""
+            previous_cap = latest_results["flow_cap"].where(spores_techs)
+            relative_cap = previous_cap / self.inputs["flow_cap_max"].where(
```
Comment (FLomb) [inline src/calliope/model.py:678]: A problem that I've never addressed properly is what happens if no cap_max is defined (i.e., it is `inf` as per the default). Normally, I'd just make sure that every tech has a cap_max, even when this is not due to obvious reasons (e.g., land constraints). For instance, batteries are typically unconstrained in our models, and I would arbitrarily assign them a cap_max that makes sense based on the order of magnitude of other things. 

To automate this, should we say that where no cap_max exists, the maximum relative potential is capped at a value that has the same order of magnitude as the highest cap_max in the model?

File: src/calliope/model.py [hunk 13]
```diff
@@ -567,3 +574,181 @@ def _recalculate_storage_initial(self, results: xr.Dataset) -> xr.DataArray:
...
+                previous_cap.fillna(0)
+                .where(previous_cap.isnull(), other=np.random.rand(*previous_cap.shape))
+                .where(spores_techs)
+            )
+
+            return new_score
+
+        def _score_evolving_average() -> xr.DataArray:
+            """Evolving average scoring algorithm."""
+            previous_cap = latest_results["flow_cap"]
```
Comment (FLomb) [inline src/calliope/model.py:703]: In all other methods, it is always: `previous_cap = latest_results["flow_cap"].where(spores_techs)`. In 'evolving_average', the `.where(spores_techs)` bit is gone. Is that on purpose?

**Review by FLomb — COMMENTED** [review]
This looks very good. Thanks a lot, @brynpickering, for converting the scoring methods from my branch based on an older Calliope version.

I have left a couple of minor comments for things that either need a double-check or could be marginally improved.

File: src/calliope/model.py (hunk 13 above)
Comment (brynpickering) [reply src/calliope/model.py:703]: It's just applied later on

Comment (FLomb) [reply src/calliope/model.py:703]: Ok, that's what I figured. I guess it saves some lines of code to do it only on the final product in this case. But could this also be potentially less efficient if you must do the averaging, subtraction and division operations for larger xarray objects? Imagine a case in which you have a humongous model, but only do spores for a couple of techs.

**Review by FLomb — COMMENTED** [review]
Thanks for the additional commits @brynpickering. I think this looks good now and can be merged!

File: src/calliope/model.py (hunk 13 above)
Comment (brynpickering) [reply src/calliope/model.py:708]: These are quite efficient operations as they're just standard arithmetic applied to known types, so shouldn't be an issue. We can revisit if it does seems slow.
